### ISSUE TITLE
Get model from Aggregate object

### ISSUE THREAD

**Issue by C-Duv** [Automattic/mongoose#7608]
**Do you want to request a *feature* or report a *bug*?**

Request a feature.

**What is the current behavior?**

Using the following code:
```Js
mongoose.set('debug', function(collectionName, method, query, queryOptions) {
    // Enrich query or queryOptions:
    query.foo = 'bar';
    queryOptions.foo = 'barOpt';
});
MySchema.post('find', postHook = function(result) {
    // Attempt to read `this.foo` or `this.options.foo`
});
```
Any properties added to `query` or `queryOptions` (in `debug`-hook) won't be available to `post`-hook.

Note that enrichment done in `pre`-hook is available to `post`-hook (and to `debug`-hook for some queries, see #7606).

Here is a full gist to reproduce:
https://gist.github.com/C-Duv/543f4be151aa3ae42499a09536d56571

**What is the expected change?**

Some way to enrich the query or pass information from `debug`-hook to `post`-hook

**What is your final use case?**

I want to log, for every request, 
* The collection name (easily available in `debug`-hook, more difficult in `post`-hook)
* The method name (easily available in `debug`-hook, more difficult in `post`-hook)
* The start time (via a `pre`-hook)
* The end time (via `post`-hook)
* The actual query (from `post`-hook)
* Number of document found by the query (from `post`-hook)

**Please mention your node.js, mongoose and MongoDB version.**

* NodeJS: v10.15.3
* Mongoose: v5.4.19
* MongoDB: v4.0.3

**Comment by C-Duv** [comment]
I guess, the `if (debug)` block of `iter()` function in `lib/drivers/node-mongodb-native/collection.js`
https://github.com/Automattic/mongoose/blob/944afb6130efa0bc84e3fc592b7e0785555b3c73/lib/drivers/node-mongodb-native/collection.js#L134-L143
could be adapted to exploit whatever the `debug` callback function returns (line 136) and add it to some object.

I tried something like:
```Patch
--- lib/drivers/node-mongodb-native/collection.js	2019-03-15 12:19:03.339991790 +0100
+++ lib/drivers/node-mongodb-native/collection.js	2019-03-15 12:19:02.335983086 +0100
@@ -133,7 +133,7 @@
 
     if (debug) {
       if (typeof debug === 'function') {
-        debug.apply(_this,
+        args[0].__debug = debug.apply(_this,
           [_this.name, i].concat(sliced(args, 0, args.length - 1)));
       } else if (debug instanceof stream.Writable) {
         this.$printToStream(_this.name, i, args, debug);
```
But it seems very ugly and does not work with all queries (find, create, delete, etc.).

**Comment by C-Duv** [comment]
Hello @vkarpov15, I see you added a milestone to this feature request: I'm glad you find it useful enough to be added :)

If you need anything to achieve it (tester, etc.) don't hesitate to ask me, I would happily try to help you.

**Comment by vkarpov15** [comment]
@C-Duv I'm not sure I understand what the issue is here. It looks like the debug output that your script prints is correct:

```
[2019-04-03T23:20:35.934Z] DEBUG: post-hook: query is Aggregate
[2019-04-03T23:20:35.934Z] DEBUG: post-hook: Read this._preHookVar:  fooFromPre
[2019-04-03T23:20:35.934Z] DEBUG: post-hook: Read this.options._preHookVar:  fooFromPre-viaOptions
[2019-04-03T23:20:35.940Z] DEBUG: post-hook: query is Query
[2019-04-03T23:20:35.940Z] DEBUG: post-hook: Read this._preHookVar:  fooFromPre
[2019-04-03T23:20:35.940Z] DEBUG: post-hook: Read this.options._preHookVar:  fooFromPre-viaOptions
```

It looks like both `_preHookVar` variables are getting set. Can you please clarify what the issue is here? I might be missing something.

**Comment by C-Duv** [comment]
Yes, variables set in `pre`-hook are available in post`-hook just fine (`pre` to `post` is OK).

What's missing are `fooFromDebug` and `fooFromDebug-viaOptions` values.
I set thoses to `query._debugHookVar ` and `queryOptions._debugHookVar ` respectively in the `debug`-hook (lines 17 and 19 of my reproduction script) but they are not available when inside the `post`-hook.

**Comment by vkarpov15** [comment]
So I looked more closely at this and here's how you access data from the `debug` function in your post hook:

```javascript
const postHook = function(result) {
    console.log(`[${new Date().toISOString()}] DEBUG: post-hook: query is ${this.constructor.name}`);

    // The "query" in the debug function may come from either a Query or an Aggregate, so it might end up in a different place
    const _debugHookVar = this instanceof mongoose.Query ? this.getQuery()._debugHookVar : this.pipeline()._debugHookVar;
    console.log(`[${new Date().toISOString()}] DEBUG: post-hook: Read this._debugHookVar: `, _debugHookVar);

    if (this.hasOwnProperty('_preHookVar')) {
        console.log(`[${new Date().toISOString()}] DEBUG: post-hook: Read this._preHookVar: `, this._preHookVar);
    }
    if (
        this.hasOwnProperty('options')
        && this.options.hasOwnProperty('_preHookVar')
    ) {
        console.log(`[${new Date().toISOString()}] DEBUG: post-hook: Read this.options._preHookVar: `, this.options._preHookVar);
    }

    this._postHookVar = 'fooFromPost';
    if (this.hasOwnProperty('options')) {
        this.options._postHookVar = 'fooFromPost-viaOptions';
    }
}
```

Unfortunately the `fooFromDebug-viaOptions` will not work because Mongoose clones options before sending them to the `debug` function, for example [here](https://github.com/Automattic/mongoose/blob/5bf17b861d927fd402d56e0a38cfa60db8325783/lib/aggregate.js#L943).

Going back to your original question, you want to get the collection name and operation in your post hook. That's much easier than trying to get something useful out of the `debug` function, which is at a very different abstraction layer than pre and post hooks. Here's how you'd print the current operation and collection name in a post hook:

```javascript
const postHook = function(result) {
    console.log(`[${new Date().toISOString()}] DEBUG: post-hook: query is ${this.constructor.name}`);

    require('assert').ok(this instanceof mongoose.Query || this instanceof mongoose.Aggregate);
    const model = this instanceof mongoose.Query ? this.model : this._model;
    const op = this instanceof mongoose.Query ? this.op : 'aggregate';
    console.log(`${model.collection.collectionName}.${op}()`);

    if (this.hasOwnProperty('_preHookVar')) {
        console.log(`[${new Date().toISOString()}] DEBUG: post-hook: Read this._preHookVar: `, this._preHookVar);
    }
    if (
        this.hasOwnProperty('options')
        && this.options.hasOwnProperty('_preHookVar')
    ) {
        console.log(`[${new Date().toISOString()}] DEBUG: post-hook: Read this.options._preHookVar: `, this.options._preHookVar);
    }

    this._postHookVar = 'fooFromPost';
    if (this.hasOwnProperty('options')) {
        this.options._postHookVar = 'fooFromPost-viaOptions';
    }
}
```

**Comment by vkarpov15** [comment]
I found a related problem looking at this issue though: there's no way to get a reference to the current aggregation's model in a post hook without doing `this._model`, which is undocumented. There's the [`Aggregate#model()` function](https://mongoosejs.com/docs/api.html#aggregate_Aggregate-model), but that's a setter only, it doesn't actually get the model.

**Comment by C-Duv** [comment]
Thanks for the `postHook`, exactly what I was looking for (I had already tried to read some properties but was not sure which I could read without too much danger):

I also added the following code to get the original request:
```
const request = this instanceof mongoose.Query ? this._conditions : this._pipeline;
// Will return:
// * For find() : { field1: 2 }
// * For aggregate() : [ { '$project': { a: 1, b: 1 } }, { '$limit': 1 } ]
```
(request examples are taken from [my original gist](https://gist.github.com/C-Duv/543f4be151aa3ae42499a09536d56571))

Also, with 0ea126d (available since v5.5.1),  the following line of your `postHook` function:
```
const model = this instanceof mongoose.Query ? this.model : this._model;
```
Can be replaced by a simple:
```
const model =  this..model();
```
Right?

**Comment by vkarpov15** [comment]
@C-Duv yep. Before v5.5.1 there wasn't a way to get an Aggregate object's model without doing `this._model`, which technically isn't part of the public API.
